File: python_cpp/uPIMulator_frontend/src/iss/sram/wram.py
from typing import List

from abi.word.data_address_word import DataAddressWord
from abi.word.data_word import DataWord
from abi.word.representation import Representation
from util.config_loader import ConfigLoader


class WRAM:
    def __init__(self):
        self._address: DataAddressWord = DataAddressWord()
        self._address.set_value(ConfigLoader.wram_offset())

        self._data_words: List[DataWord] = [DataWord() for _ in range(ConfigLoader.wram_size() // DataWord().size())]

    def address(self) -> int:
        return self._address.value(Representation.UNSIGNED)

    def read(self, address: int) -> DataWord:
        index = self._index(address)
        data_word = DataWord()
        data_word.set_value(self._data_words[index].value(Representation.UNSIGNED))
        return data_word

    def write(self, address: int, data_word: DataWord) -> None:
        index = self._index(address)
        self._data_words[index].set_value(data_word.value(Representation.UNSIGNED))

    def _index(self, address: int) -> int:
        assert address % DataWord().size() == 0
        index = (address - self.address()) // DataWord().size()
        assert 0 <= index < len(self._data_words)

        return index

File: python_cpp/uPIMulator_backend/src/simulator/sram/wram.cc
#include "simulator/sram/wram.h"

namespace upmem_sim::simulator::sram {

WRAM::WRAM() {
  address_ = new abi::word::DataAddressWord();
  address_->set_value(util::ConfigLoader::wram_offset());

  size_ = util::ConfigLoader::wram_size();

  assert(address() % abi::word::DataWord().size() == 0);
  assert(size_ % abi::word::DataWord().size() == 0);

  cells_.resize(num_data_words());
  for (int i = 0; i < num_data_words(); i++) {
    cells_[i] = new abi::word::DataWord();
  }
}

WRAM::~WRAM() {
  delete address_;

  for (int i = 0; i < num_data_words(); i++) {
    delete cells_[i];
  }
}

int64_t WRAM::read(Address address) {
  return cells_[index(address)]->value(abi::word::UNSIGNED);
}

void WRAM::write(Address address, int64_t value) {
  cells_[index(address)]->set_value(value);
}

void WRAM::write(Address address, encoder::ByteStream *byte_stream) {
  cells_[index(address)]->from_byte_stream(byte_stream);
}

int WRAM::index(Address address) {
  assert(address >= this->address());
  assert(address + abi::word::DataWord().size() <= this->address() + size_);
  assert((address - this->address()) % abi::word::DataWord().size() == 0);

  return static_cast<int>((address - this->address()) /
                          abi::word::DataWord().size());
}

}  // namespace upmem_sim::simulator::sram